impl/monitor/monitor.go

```diff
@@ -0,0 +1,247 @@
+func (s *Server) GetSignedMapRootStream(in *ktpb.GetMonitoringRequest, stream mspb.MonitorService_GetSignedMapRootStreamServer) error {
+	// TODO(ismail): implement stream API
+	return grpc.Errorf(codes.Unimplemented, "GetSignedMapRootStream is unimplemented")
+}
+
+// GetSignedMapRootByRevision works similar to GetSignedMapRoot but returns
+// the monitor's result for a specific map revision.
+//
+// Returns the signed map root for the specified epoch the monitor observed.
+// If the monitor could not reconstruct the map root given the set of
+// mutations from the previous to the current epoch it won't sign the map root
+// and additional data will be provided to reproduce the failure.
+func (s *Server) GetSignedMapRootByRevision(ctx context.Context, in *ktpb.GetMonitoringRequest) (*ktpb.GetMonitoringResponse, error) {
+	// TODO(ismail): implement by revision API
+	return nil, grpc.Errorf(codes.Unimplemented, "GetSignedMapRoot is unimplemented")
+}
+
+func (s *Server) pollMutations(ctx context.Context, opts ...grpc.CallOption) ([]*ktpb.Mutation, error) {
+	req := &ktpb.GetMutationsRequest{PageSize: pageSize, Epoch: s.nextRevToQuery()}
+	resp, err := s.client.GetMutations(ctx, req, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	seen := time.Now().UnixNano()
+	if got, want := resp.GetSmr(), s.lastSeenSMR(); bytes.Equal(got.GetRootHash(), want.GetRootHash()) &&
+		got.GetMapRevision() == want.GetMapRevision() {
+		// We already processed this SMR. Do not update seen SMRs. Do not scroll
+		// pages for further mutations. Return empty mutations list.
+		glog.Infof("Already processed this SMR with revision %v. Continuing.", got.GetMapRevision())
+		return nil, nil
+	}
+
+	mutations := make([]*ktpb.Mutation, pageSize*2)
+	mutations = append(mutations, resp.GetMutations()...)
+	if err := s.pageMutations(ctx, resp, mutations, opts...); err != nil {
+		glog.Errorf("s.pageMutations(_): %v", err)
+		return nil, err
+	}
+
+	// update seen SMRs:
+	// TODO: this should be
+
+	rh := resp.GetSmr().GetRootHash()
+	switch err := s.verifyMutations(mutations, rh); err {
+	// TODO(ismail): return proper data for failure cases:
+	case ErrInvalidMutation:
+		glog.Errorf("TODO: handle this ErrInvalidMutation properly")
+	case ErrInvalidSignature:
+		glog.Errorf("TODO: handle this ErrInvalidSignature properly (return SMR)")
+	case ErrNotMatchingRoot:
+		glog.Errorf("TODO: handle this ErrNotMatchingRoot properly")
+	case nil:
+		glog.Info("TODO: Successfully verified.")
+	default:
+		glog.Errorf("Unexpected error: %v", err)
+	}
+
+	// TODO(ismail): Sign reconstructed Hash instead of response hash:
+	sig, err := s.signer.SignObject(resp.GetSmr().GetRootHash())
+	if err != nil {
+		return nil, fmt.Errorf("s.signer.SignObject(_): %v", err)
+	}
+	// Update seen/processed signed map roots:
+	s.proccessedSMRs = append(s.proccessedSMRs,
+		&ktpb.GetMonitoringResponse{
+			Smr: &trillian.SignedMapRoot{
```

Comment:
As discussed there will be 2 PRs. One monitor that doesn't do anything and just signs what it sees and a separate pull for the verification logic.